File: service/notification-service/src/main/java/com/main_project/notification_service/controller/NotificationController.java
package com.main_project.notification_service.controller;

import com.main_project.notification_service.dto.NotificationDTO;
import com.main_project.notification_service.entity.Notification;
import com.do_an.common.event.InvoicePaidNotificationEvent;
import com.main_project.notification_service.dto.request.InvoicePaidRequest;
import com.main_project.notification_service.repository.NotificationRepository;
import com.main_project.notification_service.service.WebSocketNotificationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.axonframework.eventhandling.EventBus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import static org.axonframework.eventhandling.GenericEventMessage.asEventMessage;



@RestController
@RequestMapping("/notification-service/notifications")
@RequiredArgsConstructor
@Slf4j
public class NotificationController {
    private final EventBus eventBus;
    private final WebSocketNotificationService webSocketNotificationService;
    private final NotificationRepository notificationRepository;

    @PostMapping("/test/invoice-paid")
    public ResponseEntity<Map<String, Object>> simulateInvoicePaid(
            @RequestBody(required = false) InvoicePaidRequest request) {

        log.info("🔔 [TEST API] Simulating InvoicePaidNotificationEvent");


        // Tạo event
        InvoicePaidNotificationEvent event = new InvoicePaidNotificationEvent(
                UUID.fromString(request.getInvoiceId()),
                UUID.fromString(request.getAppointmentId()),
                "Hóa đơn " + request.getInvoiceId() + " đã được thanh toán thành công. Có thể cấp phát đơn thuốc."
        );

        // Phát event qua Axon EventBus
        eventBus.publish(asEventMessage(event));

        log.info("✅ [TEST API] InvoicePaidNotificationEvent published: invoiceId={}, appointmentId={}",
                request.getInvoiceId(), request.getAppointmentId());

        // Trả về response
        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("message", "InvoicePaidNotificationEvent đã được phát thành công");

        return ResponseEntity.ok(response);
    }


    @GetMapping("/user/{userId}")
    public ResponseEntity<List<NotificationDTO>> getNotificationsByUserId(@PathVariable UUID userId) {
        log.info("Getting notifications for userId: {}", userId);
        List<Notification> notifications = notificationRepository.findByUserIdOrderByCreatedAtDesc(userId);
        log.info("Found {} notifications for userId: {}", notifications.size(), userId);
        List<NotificationDTO> dtos = notifications.stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
        return ResponseEntity.ok(dtos);
    }

    @GetMapping("/user/{userId}/unread-count")
    public ResponseEntity<Long> getUnreadCount(@PathVariable UUID userId) {
        log.info("Getting unread count for userId: {}", userId);
        long count = notificationRepository.countByUserIdAndStatus(userId, "sent");
        log.info("Unread count for userId {}: {}", userId, count);
        return ResponseEntity.ok(count);
    }

    @GetMapping("/template")
    public ResponseEntity<List<NotificationDTO>> getNotificationsByTemplateId(@RequestParam String templateId) {
        log.info("Getting notifications for templateId: {}", templateId);
        List<Notification> notifications = notificationRepository.findByTemplateIdOrderByCreatedAtDesc(templateId);
        log.info("Found {} notifications for templateId: {}", notifications.size(), templateId);
        List<NotificationDTO> dtos = notifications.stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
        return ResponseEntity.ok(dtos);
    }

    @PutMapping("/{id}/read")
    public ResponseEntity<Void> markAsRead(@PathVariable UUID id) {
        notificationRepository.findById(id).ifPresent(notification -> {
            notification.setStatus("read");
            notificationRepository.save(notification);
        });
        return ResponseEntity.ok().build();
    }

    private NotificationDTO toDTO(Notification notification) {
        return NotificationDTO.builder()
                .id(notification.getId())
                .userId(notification.getUserId())
                .channel(notification.getChannel())
                .templateId(notification.getTemplateId())
                .message(notification.getMessage())
                .status(notification.getStatus())
                .errorMessage(notification.getErrorMessage())
                .retryCount(notification.getRetryCount())
                .createdAt(notification.getCreatedAt())
                .updatedAt(notification.getUpdatedAt())
                .build();
                
    }
}

File: fe/src/controllers/NotificationController.ts
import { API_CONFIG, createApiUrl, getApiHeaders } from "../config/api";

export interface NotificationDTO {
  id: string;
  userId: string;
  channel: string;
  templateId: string;
  message: string;
  status: string;
  errorMessage?: string;
  retryCount: number;
  createdAt: string;
  updatedAt: string;

  invoiceId?: string;
  dispenseOrderId?: string;
  appointmentId?: string;
  type?: string;
  timestamp?: number;


}

class NotificationController {
  private baseUrl = API_CONFIG.ENDPOINTS.NOTIFICATIONS;

  private async handleResponse<T>(res: Response): Promise<T> {
    if (!res.ok) {
      let msg = `HTTP Error: ${res.status}`;
      try {
        const data = await res.json();
        if (data?.message) msg = data.message;
      } catch (error) {
        // ignore parse error
      }
      throw new Error(msg);
    }
    return res.json();
  }

  async getByUserId(userId: string): Promise<NotificationDTO[]> {
    const res = await fetch(createApiUrl(this.baseUrl, `/user/${userId}`), {
      headers: getApiHeaders(true),
    });
    return this.handleResponse<NotificationDTO[]>(res);
  }

  async getByTemplateId(templateId: string): Promise<NotificationDTO[]> {
    const encodedTemplateId = encodeURIComponent(templateId);
    const res = await fetch(createApiUrl(this.baseUrl, `/template?templateId=${encodedTemplateId}`), {
      headers: getApiHeaders(true),
    });
    return this.handleResponse<NotificationDTO[]>(res);
  }

  async getUnreadCount(userId: string): Promise<number> {
    const res = await fetch(createApiUrl(this.baseUrl, `/user/${userId}/unread-count`), {
      headers: getApiHeaders(true),
    });
    return this.handleResponse<number>(res);
  }

  async markAsRead(notificationId: string): Promise<void> {
    const res = await fetch(createApiUrl(this.baseUrl, `/${notificationId}/read`), {
      method: 'PUT',
      headers: getApiHeaders(true),
    });
    await this.handleResponse<void>(res);
  }
}

export const notificationController = new NotificationController();